src/Token/Builder.php

```diff
@@ -16,6 +16,7 @@
 use function array_merge;
 use function in_array;
 
+/** @immutable */
```

Comment:
The interface should probably also be immutable

Reply:
This idea puzzled me for the past hour.

Of course it must be marked `@immutable` as well, otherwise we could end with different implementations and that could be catastrophic for the consumer.

But, on the other hand, what sense does it make to mark the "*statefulness*" for an interface?

:thinking:

Reply:
This is a bit of a quirk of phpstan vs psalm.

In psalm, immutable means "doesn't change internal state", plus "no side-effects" :) 

As a consumer of an immutable interface, you know that calling methods on it won't lead to an internal state mutation, which is something that PHP doesn't enforce natively, even with PHP 8.2 readonly classes.

In practice, `@immutable` is expected on both levels, not just in the interface: in fact, psalm will even report this as an issue.

Reply:
Here's an example of why immutability is important from a consumer/contract PoV

Reply:
Note: I re-used this example to state my point **against** the current state of https://wiki.php.net/rfc/readonly_amendments

Reply:
My understanding is that the RFC wouldn't allow making the `count` property mutable in a child of a readonly class. I think they're just meaning to allow a mutable child of an readonly class to declare **additional** properties, which would be mutable. Not change the ones that are already there. Seems reasonable to me.

Reply:
@bdsl the properties are completely irrelevant: the behavior changes based on the public API of the object, which is no longer immutable.

I suggest you discuss this further in https://externals.io/message/119007#119016 - here I only explained why the immutability has to be declared at interface level, and respected in all implementations.

Reply:
I wasn't able to get signup for that mailing list working with my email last time I tried, so I'll leave it for now.